src/main/java/org/xbill/DNS/lookup/LookupSession.java

```diff
@@ -0,0 +1,295 @@
+      return completedFuture(new LookupResult(records, null));
+    }
+    return null;
+  }
+
+  private <T extends LookupFailedException> CompletionStage<LookupResult> completeExceptionally(
+      T failure) {
+    CompletableFuture<LookupResult> future = new CompletableFuture<>();
+    future.completeExceptionally(failure);
+    return future;
+  }
+
+  private CompletionStage<LookupResult> resolveRedirects(LookupResult response, Record query) {
+    CompletableFuture<LookupResult> future = new CompletableFuture<>();
+    maybeFollowRedirect(response, query, 1, future);
+    return future;
+  }
+
+  private void maybeFollowRedirect(
+      LookupResult response,
+      Record query,
+      int redirectCount,
+      CompletableFuture<LookupResult> future) {
+    try {
+      if (redirectCount > maxRedirects) {
+        throw new RedirectOverflowException(
+            format("Refusing to follow more than %s redirects", maxRedirects));
+      }
+
+      List<Record> records = response.getRecords();
+      if (records.isEmpty()) {
+        future.complete(response);
+      } else if (records.get(0).getType() == DNAME || records.get(0).getType() == CNAME) {
+        lookupWithCache(
+                buildRedirectQuery(response, query),
+                makeAliases(response.getAliases(), query.getName()))
+            .thenAccept(m -> maybeFollowRedirect(m, query, redirectCount + 1, future));
+      } else {
+        future.complete(response);
+      }
+    } catch (LookupFailedException e) {
+      future.completeExceptionally(e);
+    }
+  }
+
+  /** Return an unmodifiable list containing the contents of previous, if any, plus name. */
+  private List<Name> makeAliases(List<Name> previous, Name name) {
+    if (previous == null) {
+      return singletonList(name);
+    } else {
+      List<Name> copy = new ArrayList<>(previous);
+      copy.add(name);
+      return Collections.unmodifiableList(copy);
+    }
+  }
+
+  private Record buildRedirectQuery(LookupResult response, Record question) {
+    List<Record> answer = response.getRecords();
+    Record firstAnswer = answer.get(0);
+    if (answer.size() != 1) {
+      throw new InvalidZoneDataException("Multiple CNAME RRs not allowed, SEE RFC1034 3.6.2");
+    }
+
+    if (firstAnswer.getType() == CNAME) {
+      return Record.newRecord(
+          ((CNAMERecord) firstAnswer).getTarget(), question.getType(), question.getDClass());
+    }
+    // if it is not a CNAME, it's a DNAME
+    try {
+      Name name = question.getName().fromDNAME((DNAMERecord) firstAnswer);
+      return Record.newRecord(name, question.getType(), question.getDClass());
+    } catch (NameTooLongException e) {
+      throw new InvalidZoneDataException(
+          "DNAME redirect would result in a name that would be too long");
+    }
+  }
+
+  /** Returns a LookupResult if this response was a non-exceptional empty result, else null. */
+  private static LookupResult buildResult(Message answer, List<Name> aliases) {
```

Comment:
Consider if you really need to throw the exceptions or if you could just return a failed future (the stacktrace in the exceptions is useless and it would probably half your exception-throwing-testbench-time).

And it could be useful to add the query's Name/Type to the exception; the caller should know it, but if the failure/exception is blindly used for logging it's helpful.

Reply:
One could move away from having regular synchronous flow in these methods and instead allow for a mechanism that returned either a value or an exception, but I think that might be the wrong tradeoff as it diverges from a very well established pattern of execution with a simple return value for the successful case and an Exception for the exceptional case, simply to save a ms or two. 

I'm not so sure about adding other potentially useful information to the exception that is also available up the call stack. It makes things a bit simpler to not add it at the very least :)

Reply:
setResponseToMessageFuture(...) does such a non-throwing, failed-future flow. But I agree that this is currently more readable.

There were a few issues that requested adding such information to exceptions.